### ISSUE TITLE
yolov8 classification training, resume=True not working . 

### ISSUE THREAD

**Issue by Irtiza17** [ultralytics/ultralytics#4428]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Image 1:
![image](https://github.com/ultralytics/ultralytics/assets/107918343/392a0155-7e5d-4552-8960-9f02d381bc9d)

My yolov8 classification training crashed after 20 epochs. Total epochs were 100. I saved the last checkpoints from /content/runs/classify/train8/weights/last.pt . Then i loaded the model again , with the argument resume =  True, but it started again from epoch1, as can be seen from the image provided.

Image 2:
![image](https://github.com/ultralytics/ultralytics/assets/107918343/f851fb52-4acc-4533-ab79-785e28974851)

Also the 2nd image, shows that even though i gave the resume=True, the resume still appears as False in the console logs.

### Environment

Training on Google Colab

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Irtiza17 thank you for the bug report!

Yes, this is a known issue. 

The classification trainer is much simpler than our other tasks and does not save resume-capable checkpoints that would require attaching EMAs and optimizer gradients. I'll add a TODO to work on this, but it is not a quick fix so it may take time to introduce this feature. 

In the meantime you can always start training a new model from your crashed model. This will benefit from the partial pretraining but will initiate the LR scheduler from 0 as you've seen, i.e. a 'new' training.

https://github.com/ultralytics/ultralytics/blob/162e4035eb5dd2d1f5adcc729eea738f179f4c33/ultralytics/models/yolo/classify/train.py#L64

**Comment by Irtiza17** [comment]
Thank you for the swift response

**Comment by Hir98** [comment]
@Irtiza17 I want your help.
As you have mentioned that you have store weight of model after 20 epochs.and youhave trained model upto 100 epoch.
same i want to train model upto 50 model and i want to store weight of model after every 10 epoch.
so how can i do like this??
can you please help me.

**Comment by AshishRaghani23** [comment]
@Hir98 Hello, you can pass argument save_period=10 to save weights for every 10 epochs.

**Comment by glenn-jocher** [comment]
@AshishRaghani23 Hello! You can definitely achieve this by modifying the save_period argument in your training setup. 

When you initiate a training session, there is a parameter called save_period. By default, the save_period is set to the maximum number of epochs you have defined for your training. This means that the model saves the weights at the end of the training.

However, you can specify that you want to save the weights every 10 epochs by setting save_period=10. This way, your model will automatically save the weights after every 10 epochs during the training process.

I hope this helps! Feel free to reach out if you have any other questions or issues.

**Comment by glenn-jocher** [comment]
@Hir98 @AshishRaghani23 @Irtiza17 great news 😃! Your original issue may now be resolved ✅ in PR #4482. You can now resume Classify models following the examples in the Docs at https://docs.ultralytics.com/modes/train/#resuming-interrupted-trainings

<img width="1413" alt="Screenshot 2023-08-21 at 20 31 39" src="https://github.com/ultralytics/ultralytics/assets/26833433/3c88929b-3aff-4611-afab-e2374b609774">

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by AshishRaghani23** [comment]
@glenn-jocher Great work, Thank you.

**Comment by Irtiza17** [comment]
@glenn-jocher Thank you so much!

**Comment by glenn-jocher** [comment]
@AshishRaghani23 you're welcome! I'm glad I could help. Don't hesitate to reach out if you have any more questions or issues. Happy training with YOLOv8!

**Comment by Hir98** [comment]
@glenn-jocher I am not able to train model when i pass save_period=10 during training.
![image](https://github.com/ultralytics/ultralytics/assets/47842672/d44d9d4f-c01b-4aa1-965f-a3d80f8127b2)

**Comment by Amir-Nassimi** [comment]
hello @glenn-jocher and many thanks for developing Ultralytics YOLOv8.0.208.
I have found that in the ultralytics/engine/trainer.py you have intentionally initialized the self.resume=False. therefore, even when I tried to set the resume to True, the model wont recognized it.
the question is: should set that to true manually whenever I wanted to resume a training? a detection task.

![Screenshot 2023-11-09 143830](https://github.com/ultralytics/ultralytics/assets/100153039/3be739b2-2209-4b34-aa0d-8095d1bb2dbf)

**Comment by glenn-jocher** [comment]
@Amir-Nassimi hello,

Thank you for reaching out with your question regarding the resumption of training in YOLOv8. I appreciate your attention to detail in examining the trainer.py file.

The `self.resume` variable within the trainer script is indeed set to `False` by default. This is set up primarily for safety to prevent unintended resumption of training sessions. However, when you want to resume training from a checkpoint, you do not need to modify the codebase manually.

Instead, you should pass the appropriate argument to the training script to enable the resumption of training. The script is designed to accept a parameter that sets this variable to `True`. When you pass the correct argument, the training session will recognize it and continue from the last saved checkpoint.

Make sure that the argument is correctly specified in the command you use to launch the training session. If you've done this and the parameter doesn't seem to be taking effect, it could be due to an oversight or a bug, and I would recommend reviewing the latest documentation and notes for any changes in how the resumption of training is handled in the current version of YOLOv8.

If the problem persists, please provide additional details about the commands used and any console logs that illustrate the issue, and our team will be happy to take a closer look to assist you further.

Best regards and happy detecting with YOLOv8!

**Comment by Kehoe1** [comment]
Hello, thank you very much for creating YOLOv8. However, I have a question now as follows:
I have set 100 epochs and have completed the training for 100 epochs, resulting in the generation of last.pt and best.pt. If I want to continue training for another 100 epochs, how should I proceed? Should I use the last.pt file directly as the weight file?

**Comment by Amir-Nassimi** [comment]
> Hello, thank you very much for creating YOLOv8. However, I have a question now as follows:
> I have set 100 epochs and have completed the training for 100 epochs, resulting in the generation of last.pt and best.pt. If I want to continue training for another 100 epochs, how should I proceed? Should I use the last.pt file directly as the weight file?

Hi
I've tried this before. Since you actually have trained your neural network completely, there is nothing to resume since the neural net has already been trained as much as the specific epochs you set before. If you want to re-train the network, you should consider it as a new training stage from beginning, but have the last.py as the pretrained weights of your net.
Hope this helps

**Comment by glenn-jocher** [comment]
@Amir-Nassimi hello!

To continue training your YOLOv8 model for an additional 100 epochs using the weights you've obtained from the previous training session, you can start a new training process with the `last.pt` file as your initial weights.

When starting the additional training, you'd simply specify the path to your `last.pt` file with the appropriate argument to load these weights at the beginning of the training command. This will initialize your model with the weights you've already trained, allowing further improvement upon them. Keep in mind that the learning rate scheduler and other training parameters will start anew unless specifically designed to resume from the point they left off.

If you have any other questions or need further assistance, our team is always ready to assist.

**Comment by alexandrewillame** [comment]
Hello,

I do beleive this bug is not fixed / have reappeared, as if I try to train any algorithm, the resume parameter is always at False, even if I set it to True in the model.train parameters. Tested with YOLOv8.0.222:

Ultralytics YOLOv8.0.222 🚀 Python-3.11.6 torch-2.1.0 CUDA:0 (Quadro RTX 5000, 16384MiB)

Code example to reproduce:

```
from ultralytics.models.yolo import YOLO
test0=YOLO("yolov8s-cls.pt")
test0.train(epochs=1) # This training is a new training and it is normal that resume=False
test0.train(resume=True, epochs=1) # This training still have resume=False as parameters whereas it should be resume=True. This is not normal
```

![image](https://github.com/ultralytics/ultralytics/assets/7422936/540ba265-d90b-4434-9df1-f6dca590b8da)

Environment
ultralytics.checks()
Ultralytics YOLOv8.0.222 🚀 Python-3.11.6 torch-2.1.0 CUDA:0 (GeForce GTX 1080, 8192MiB)
Setup complete ✅ (12 CPUs, 63.9 GB RAM, 191.2/476.2 GB disk)

OS: Windows 11
Environment: Spyder
Python: 3.11.6
Install:
RAM: 64GB
CPU: i7-8700K

**Comment by ManzarIMalik** [comment]
Hi @glenn-jocher ,

I am using the latest version of Ultralytics and I can second that this bug has reappeared, as I am unable to set ‘resume=True’ while training from last checkpoint, this stays to be False in params. 

I am using 
`yolo task=detect mode=train resume=True model=path/to/last.pt data=data.yaml imgsz=640 batch=8 plots=True`

This is starting training from scratch again.

**Comment by glenn-jocher** [comment]
@ManzarIMalik hi there,

We appreciate the feedback regarding the training resumption issue with `resume=True`. It seems there might be a bug affecting this feature. We'll investigate and address this in an upcoming update. Thanks for your patience and support.

Best regards.

**Comment by ManzarIMalik** [comment]
Hello @glenn-jocher,

Is there any update on resolving this issue? If not any sooner, can you please let me know what is the alternate way of resuming training from last weights using CLI please?

**Comment by glenn-jocher** [comment]
Hello @ManzarIMalik,

We are actively looking into the resumption issue. To resume training from the last weights using CLI, indicate the checkpoint file with `--weights /path/to/last.pt` and set the initial epoch with `--resume --start_epoch 100` (adjusting the starting epoch as needed).

Best regards.
